borg/repository.py

```diff
@@ -306,15 +308,44 @@ def complete_xfer():
                         segments[segment] -= 1
                     elif tag == TAG_DELETE:
                         if index_transaction_id is None or segment > index_transaction_id:
+                            # (introduced in 6425d16aa84be1eaaf88)
+                            # This is needed to avoid object un-deletion if we crash between the commit and the deletion
+                            # of old segments in complete_xfer().
+                            #
+                            # However, this only happens if the crash also affects the FS to the effect that file deletions
+                            # did not materialize consistently after journal recovery. If they always materialize in-order
+                            # then this is not a problem, because the old segment containing a deleted object would be deleted
+                            # before the segment containing the delete.
+                            #
+                            # Consider (without this block, columns are segments, rows are keys, c=commit, i=index, P=put, D=delete)
```

Comment:
can you explain / rephrase "without this block" and "rows are keys" - it is unclear what you mean.